java/src/main/kotlin/io/spine/validation/java/ConstraintViolations.kt

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2025, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.validation.java
+
+import io.spine.base.FieldPath
+import io.spine.protodata.ast.TypeName
+import io.spine.protodata.ast.qualifiedName
+import io.spine.protodata.java.ClassName
+import io.spine.protodata.java.Expression
+import io.spine.protodata.java.StringLiteral
+import io.spine.protodata.java.mapExpression
+import io.spine.protodata.java.newBuilder
+import io.spine.protodata.java.packToAny
+import io.spine.validate.ConstraintViolation
+import io.spine.validate.TemplateString
+import io.spine.validate.checkPlaceholdersHasValue
+
+/**
+ * Yields an expression that creates a new instance of [ConstraintViolation]
+ * with the given parameters.
+ *
+ * @param errorMessage The error message template string.
+ * @param declaringType The message type being validated.
+ * @param fieldPath The path to the field containing an invalid value.
+ * @param fieldValue The field value, if any.
+ */
+public fun constraintViolation(
+    errorMessage: Expression<TemplateString>,
+    declaringType: TypeName,
+    fieldPath: Expression<FieldPath>,
+    fieldValue: Expression<*>? = null
```

Comment:
Please have some details on when this parameter may be `null`. It's somewhat strange to see `null`s in the values of fields in a Proto message.